nautobot/core/celery/__init__.py

```diff
@@ -53,60 +43,74 @@ def register_task(self, task, **options):
 
 
 @signals.import_modules.connect
-def import_jobs_as_celery_tasks(sender, database_ready=True, **kwargs):
+def import_jobs(sender=None, **kwargs):
     """
-    Import system Jobs into Celery as well as Jobs from JOBS_ROOT and GIT_ROOT.
+    Import system Jobs into Nautobot as well as Jobs from JOBS_ROOT and GIT_ROOT.
+
+    Note that app-provided jobs are automatically imported at startup time via NautobotAppConfig.ready()
+    """
+    import nautobot.core.jobs  # noqa: F401
+
+    _import_jobs_from_jobs_root()
 
-    Note that app-provided Jobs are automatically imported at startup time via NautobotAppConfig.ready()
+    try:
+        _import_jobs_from_git_repositories()
+    except ProgrammingError:  # Database not ready yet, as may be the case on initial startup and migration
+        pass
+
+
+def _import_jobs_from_jobs_root():
+    """
+    (Re)import all modules in settings.JOBS_ROOT.
     """
-    logger.debug("Importing system Jobs")
-    sender.loader.import_task_module("nautobot.core.jobs")
-
-    jobs_root = settings.JOBS_ROOT
-    if jobs_root and os.path.exists(jobs_root):
-        if jobs_root not in sys.path:
-            sys.path.append(jobs_root)
-        for _, module_name, _ in pkgutil.iter_modules([jobs_root]):
-            try:
-                logger.debug("Importing Jobs from %s in JOBS_ROOT", module_name)
-                existing_module = find_spec(module_name)
-                if existing_module is not None:
-                    existing_module_path = os.path.realpath(existing_module.origin)
-                    jobs_root_path = os.path.realpath(jobs_root)
-                    if not existing_module_path.startswith(jobs_root_path):
-                        raise ImportError(
-                            f"JOBS_ROOT Jobs module {module_name} conflicts with existing module {existing_module_path}"
-                        )
-                sender.loader.import_task_module(module_name)
-            except Exception as exc:
-                logger.exception(exc)
-
-    git_root = settings.GIT_ROOT
-    if git_root and os.path.exists(git_root):
-        if git_root not in sys.path:
-            sys.path.append(git_root)
-
-        # We can't detect which Git directories we're *supposed* to auto-load Jobs from if we can't read GitRepository
-        # records from the DB, unfortunately.
-        # We work around this in JobModel.job_task to try later loading Git jobs on-the-fly if needed.
-        if database_ready:
+    if not (settings.JOBS_ROOT and os.path.isdir(settings.JOBS_ROOT)):
+        return
+
+    # Flush any previously loaded non-system, non-App Jobs
+    for job_class_path in list(registry["jobs"]):
+        if job_class_path.startswith("nautobot."):
```

Comment:
Since we're storing the jobs in the jobs registry, would it make sense to also store the job source to make this simpler?

Reply:
It'd just move the logic from here to the `register_jobs()` function, I think - there's no built-in knowledge of "job source" in the register_jobs call.

Reply:
Do we have any docs on the difference between the jobs registry and the Job model in the database? We use the jobs registry as an in memory cache if I remember our earlier conversation correctly?

Reply:
I'll see what I can add. The Job model says "A job exists or existed at some point" but doesn't store the executable code of the Job. The registry says "Here's the source code for this Job so you can actually run it".

Reply:
But yeah, the registry is being used as a cache so we don't have to re-import the job just to introspect its existence and such like that - we *do* have to always re-import it when we go to actually run it so as to be confident that we're running the right version of the job.

Reply:
Added some more doc updates in the latest commits.